packages/apollo-voyager-offline/src/conflictHandlers.ts

```diff
@@ -0,0 +1,62 @@
+import { SyncServerError } from '@aerogear/apollo-voyager-tools'
+import debug from 'debug'
+import { CONFLICT_ERROR, CONFLICT_LOGGER } from './constants'
+import { ObjectState } from './ObjectState'
+
+export const logger = debug(CONFLICT_LOGGER)
+
+/**
+ *
+ * @param currentRecord the object state that the server knows about
+ * @param client the object state that the client knows about
+ */
+export const returnToClient: ConflictResolutionHandler = (currentRecord: ConflictResolutionData, client: ConflictResolutionData): SyncServerError => {
+  logger(`Conflict detected. Server: ${currentRecord} client: ${client}`)
+  return new SyncServerError('Conflict detected.', currentRecord, CONFLICT_ERROR)
+}
+/**
+ *
+ * @param currentRecord the object state that the server knows about
+ * @param client the object state that the client knows about
+ */
+export const clientWins: ConflictResolutionHandler = (helper: ObjectState, currentRecord: ConflictResolutionData, client: ConflictResolutionData): ConflictResolutionData => {
+  return helper.next(client)
+}
+/**
+ *
+ * @param currentRecord the object state that the server knows about
+ * @param client the object state that the client knows about
+ */
+export const serverWins: ConflictResolutionHandler = (helper: ObjectState, currentRecord: ConflictResolutionData, client: ConflictResolutionData): ConflictResolutionData => {
+  logger(`Conflict detected. Server: ${currentRecord} client: ${client}`)
+  return currentRecord
+}
+/**
+ *
+ * @param currentRecord the object state that the server knows about
+ * @param client the object state that the client knows about
+ */
+export const mergeClientOntoServer: ConflictResolutionHandler = (helper: ObjectState, currentRecord: ConflictResolutionData, client: ConflictResolutionData): ConflictResolutionData => {
+  logger(`Conflict detected. Server: ${currentRecord} client: ${client}`)
+  const serverNextState = helper.next(currentRecord)
+  const newRecord = Object.assign(serverNextState, client)
```

Comment:
I think `Object.assign` is not going to be suitable because it only shallow merges the top level properties of an object.

you can try paste this into the node repl in your terminal to see what I'm talking about.

```
// Object nesting
const x = { a: { a: 1 } }
const y = { a: { b: 1 } }
const z = Object.assign(x, y)
console.log(z) // { a: { b: 1 } }
```

We would probably want proper recursive merging. The `deepmerge` module is a good fit. It can do this and it even has options for merging arrays together as well.

Check out this blogpost: https://davidwalsh.name/javascript-deep-merge

@wtrocki @danielpassos what do you guys think?

Reply:
No longer valid. I removed controversial stuff we cannot test now. We going to add things as we go and work on individual tasks.